types/src/abi.rs

```diff
@@ -258,28 +258,41 @@ impl Type {
         gen: &GatewayABIConfig,
     ) -> Result<Vec<u8>, &'static str> {
         match self {
-            Type::Address(size) => match size {
-                20 => {
-                    let res: [u8; 20] = decode_buf2val(encoded_val)?;
-                    Ok(res.encode())
-                },
-                32 => {
-                    let res: [u8; 32] = decode_buf2val(encoded_val)?;
-                    Ok(res.encode())
-                },
-                _ => Err("Unknown Address size"),
-            },
-            Type::DynamicAddress => match gen.address_length {
-                20 => {
-                    let res: [u8; 20] = decode_buf2val(encoded_val)?;
-                    Ok(res.encode())
-                },
-                32 => {
-                    let res: [u8; 32] = decode_buf2val(encoded_val)?;
-                    Ok(res.encode())
-                },
-                _ => unimplemented!(),
+            Type::Address(size) => {
+                if size != &gen.address_length {
+                    Err("Encoded address has different length than the gateway's configuration")
+                } else {
```

Comment:
If i were to optimise this, i'd add a function on the enum to try and check this before I do anything, this check would simply either pass if the length is the same as configured or be ignored if it doesn't apply (e.g: dynamic address)

This avoids any Christmas tree like stuff where we repeat the same error

Reply:
Refactored as a method on the enum as part of the new commit.